src/rewards/FlywheelGaugeRewards.sol

```diff
@@ -0,0 +1,173 @@
+        address _owner, 
+        Authority _authority,
+        uint32 _rewardsCycleLength,
+        ERC20Gauges _gaugeToken,
+        IRewardsStream _rewardsStream
+    ) Auth(_owner, _authority) {
+        rewardToken = _rewardToken;
+        flywheel = _flywheel;
+        rewardsCycleLength = _rewardsCycleLength;
+
+        // seed initial rewardsCycleEnd
+        rewardsCycleEnd = uint32(block.timestamp) / rewardsCycleLength * rewardsCycleLength;
+
+        gaugeToken = _gaugeToken;
+
+        rewardsStream = _rewardsStream;
+    }
+
+    /**
+        @notice Iterates over all live gauges and queues up the rewards for the cycle
+        @return totalQueuedForCycle the max amount of rewards to be distributed over the cycle
+        @dev critical that there is no pagination to prevent double spending. All gauges must be synced atomically, unless the gaugeToken checkpoints rewards.
+
+        GAS: Average path for `n` gauges - SLOAD * (n + 7) + SSTORE * (n + 1) + 1 token transfer
+    */
+    function queueRewardsForCycle() external requiresAuth returns (uint256 totalQueuedForCycle) {
+        // ensure new cycle has begun
+        require(block.timestamp > rewardsCycleEnd); // SLOAD
+        
+        // next cycle is always the next even divisor of the cycle length above current block timestamp.
+        rewardsCycleEnd = (uint32(block.timestamp) + rewardsCycleLength) / rewardsCycleLength * rewardsCycleLength; // SSTORE
+
+        // queue the rewards stream and sanity check the tokens were received
+        uint256 balanceBefore = rewardToken.balanceOf(address(this)); // SLOAD
+        totalQueuedForCycle = rewardsStream.getRewards();
+        require(rewardToken.balanceOf(address(this)) - balanceBefore >= totalQueuedForCycle); // SLOAD
+
+        // iterate over all gauges and update the rewards allocations
+        address[] memory gauges = gaugeToken.gauges(); // n * SLOAD
+        uint256 size = gauges.length;
+        for (uint256 i = 0; i < size; i++) {
+            ERC20 gauge = ERC20(gauges[i]);
+            
+            QueuedRewards memory queuedRewards = marketQueuedRewards[gauge]; // SLOAD
+
+            // SSTORE
+            marketQueuedRewards[gauge] = QueuedRewards({
+                priorCycleRewards: queuedRewards.priorCycleRewards + queuedRewards.currentCycleRewards,
+                currentCycleRewards: uint128(gaugeToken.calculateGaugeAllocation(address(gauge), totalQueuedForCycle)) // SLOAD * 3
+            });
+        }
+    }
+
+    /**
+     @notice calculate and transfer accrued rewards to flywheel core
+     @param market the market to accrue rewards for
+     @param lastUpdatedTimestamp the last updated time for market
+     @return accruedRewards the amount of reward tokens accrued and transferred
+
+     GAS: Average path includes 2 SLOAD, 1 warm SSTORE and 1 ERC20 transfer.
+    */
+    function getAccruedRewards(ERC20 market, uint32 lastUpdatedTimestamp) external override returns (uint256 accruedRewards) {
```

Comment:
Similar point about `get` prefix - calculating and transferring. Wonder if this syntax is common place in other DeFi code?

Reply:
I mean its used in the most common stakingRewards contract

https://github.com/Synthetixio/synthetix/blob/v2.63.0/contracts/StakingRewards.sol#L97-L105

We can always change it though, no need to anchor to the past.

Reply:
I'm actually planning on rearchitecting so that the token transfer only happens on claims